Pandas wrapper

nick-harder opened the pull request:
<!--
SPDX-FileCopyrightText: ASSUME Developers

SPDX-License-Identifier: AGPL-3.0-or-later
-->

# Pull Request

## Related Issue
Closes #[issue number] (if applicable)

## Description
As suggested by @maurerle this improves performance of all simulations by a factor of 2 up to 4, by replacing pandas actions with numpy where possible.

For this, we are using a wrapper object FastDatetimeSeries which wraps a numpy array so that we can access it using typical datetime accessors

## Changes Proposed
- Shift to using special classes of FastIndex and FastSeries
- Adjust the rest of the code to make use of this new class

## Testing
[Describe the testing you've done, including any specific test cases or scenarios]

## Checklist
Please check all applicable items:

- [ ] Code changes are sufficiently documented (docstrings, inline comments, `doc` folder updates)
- [ ] New unit tests added for new features or bug fixes
- [ ] Existing tests pass with the changes
- [ ] Reinforcement learning examples are operational (for DRL-related changes)
- [ ] Code tested with both local and Docker databases
- [ ] Code follows project style guidelines and best practices
- [ ] Changes are backwards compatible, or deprecation notices added
- [ ] New dependencies added to `pyproject.toml`
- [ ] A note for the release notes `doc/release_notes.rst` of the upcoming release is included
- [ ] Consent to release this PR's code under the GNU Affero General Public License v3.0

## Additional Notes (if applicable)
[Any additional information, concerns, or areas you want reviewers to focus on]

## Screenshots (if applicable)
[Add screenshots to demonstrate visual changes]

nick-harder commented:
@maurerle all tests pass except for one in unit_operator def test_get_actual_dispatch. I don't understand why it doesn't work, could you please take a look into it?

maurerle commented:
Somehow the other tests for RL are not yet working @nick-harder

nick-harder commented:
@maurerle thanks for the fix, I have also fixed the RL tests now

kim-mskw commented on assume/common/forecasts.py:
```diff
@@ -82,6 +84,21 @@ def get_price(self, fuel_type: str) -> pd.Series:
+        return FastSeries(index=self.index, value=value, name=name)
+
 
 class CsvForecaster(Forecaster):
```
Could that be general? So from df to fastseries dictionary? Does not need to be related to forecaster.

nick-harder replied:
replaced with a method from FastSeries itself

kim-mskw commented on assume/strategies/advanced_orders.py:
```diff
@@ -138,7 +138,7 @@ def calculate_bids(
             # Flex-bid price formulation
             if op_time <= -unit.min_down_time or op_time > 0:
-                bid_quantity_flex = max_power[start] - bid_quantity_inflex
+                bid_quantity_flex = max_pwr - bid_quantity_inflex
```
Why shorten the variable name? I feel this is not as self explanatory

nick-harder replied:
max_power is an array, we cannot reuse the name again. we would need to rename max_power to max_power_array in the first place or use a different name here

nick-harder replied:
I mean this is just one way of iterating over a list. we can of course agree to rename the lists and iterate over them and call the variables something else. Please make a suggestion what makes more sense from your point of view

nick-harder replied:
I have changed the naming everywhere to max_power_values and max power in the iteration. this should avoid any confusions

kim-mskw commented:
I do not find it too confusing or inconvenient; it looks rather arbitrary. However, contrary to what @nick-harder  said in the development meeting, I would change all the Args Docsting to explicitly state that it is a fast series. If I encounter unexpected behaviour because I think it is a series, the information that it is not should be as close to me as possible.

kim-mskw commented on assume/common/forecasts.py:
```diff
@@ -82,6 +84,21 @@ def get_price(self, fuel_type: str) -> pd.Series:
 
         return self[f"fuel_price_{fuel_type}"]
 
+    def _to_fast_series(self, value, name: str) -> FastSeries:
```
No, I meant to ask if they could be in utils. I see that this is currently just used in the forecaster. Yet this function seems very general, and it could be helpful in various locations in the future. If it is not, please indicate why.

nick-harder commented:
@maurerle I am working on tests and updates, please wait

nick-harder commented on assume/common/units_operator.py:
```diff
@@ -248,7 +248,9 @@ def handle_data_request(self, content: DataRequestMessage, meta: MetaDict) -> No
         data = []
         try:
-            data = self.units[unit].outputs[metric_type][start:end]
+            data = (
+                self.units[unit].outputs[metric_type].as_pd_series(start=start, end=end)
+            )
```
this is an internal process of ASSUME, so hopefully this is fine like this since changing it would need to call other functions from FAstSeries, which wouldn't make the situation clearer

nick-harder commented:
@kim-mskw @maurerle I have tested all non learning examples and everythings works fine, all data is also being properly written to the database and other outputs. I have also replaced and removed pandas from the code as much as possible and renamed the typing hints and other things. Please take another look at the commit

nick-harder commented:
@maurerle I have been testing now with the largest example we have for 2019 with two EOM and CRM markets, and the speed up is crazy, I didn't expect that. The difference is around 4.2 times. We should really all put a bit more effort and merge this as soon as we can after proper testing, this performance increase is even more what we expected...

running on pandas_wrapper
![telegram-cloud-document-2-5332300261325037747](https://github.com/user-attachments/assets/2f4c93ee-ff46-473c-a953-f9e7968da672)

running on main
![telegram-cloud-document-2-5332300261325038181](https://github.com/user-attachments/assets/15fa7ab7-39b3-4a90-b0e4-eb54881709d4)

nick-harder commented:
@kim-mskw I have tested the RL examples we have. Some things had to be fixed byt everything is working and the learning in small 1 and 2 is also like 2 to 3 times faster

kim-mskw commented:
> the review! @kim-mskw would you have time to take another look at the PR if everything is good from your side and looks fine?

Of course :) on it!

kim-mskw commented on assume/scenario/loader_csv.py:
```diff
@@ -563,14 +564,14 @@ def setup_world(
 
-    save_frequency_hours = config.get("save_frequency_hours", 48)
+    # save every thousand steps by default to free up memory
+    save_frequency_hours = config.get("save_frequency_hours", 1000)
```
Did we check the writing of dfs in the output role with a high save_frequency?

nick-harder replied:
@maurerle thanks for the analysis! But 350mb compared to 500 mb, so 150mb extra is a low price to price for a significant speed up. In my opinion this change to longer periods is justified. I don't think any more or less modern system would run out of memory because of this.

maurerle replied:
On my machine it does not give a significant speed up by reducing this param. How much is it on yours?
Typically, the save_dfs function should not block anyway.

The problem is, that this increases linearly, I just stopped the simulation after some time.

nick-harder replied:
@maurerle try activating the saving to csv files, this will change the timing for you

nick-harder replied:
@maurerle also as of now on pandas_wrapper the config for example_03 has the save frequency as null, so it is never saved

maurerle replied:
@nick-harder CSV does only save at the end as `save_frequency_hours is disabled due to CSV export being enabled. Data will be stored in the CSV files at the end of the simulation.`

Which gives above graphs as well when setting `save_frequency_hours: 96` in the yaml.
So I wonder how you see a speedup with this setting

maurerle replied:
Which we did as `unit_dispatch` changes in the CSV files when the accepted_* fields are added for example.
so eventually we should fix this as I do not really like to have linearly increasing memory consumption for a simulation tool :)

Saving csv is indeed a little slower, when the above mentioned handler to switch back to None is removed.

maurerle commented on tests/test_units_operator.py:
```diff
@@ -243,24 +244,28 @@ async def test_get_actual_dispatch(units_operator: UnitsOperator):
+    assert datetime2timestamp(unit_dfs[0]["time"][0]) == last
+    assert datetime2timestamp(unit_dfs[0]["time"][1]) == clock.time
+    assert len(unit_dfs[0]["time"]) == 2
     assert len(market_dispatch) == 0
```
Why did the tests change here to write two timestamps?

We should have seen this, as this caused #539
